### DISCUSSION TITLE
We use a lot of static queries, which result in a lot of network requests

### DISCUSSION THREAD

**Discussion by turbek** [gatsbyjs/gatsby#28654] [Help]
Hi,

This is something I can't really wrap my head around.

There are two ways of creating graphql queries. Page and static queries. The main difference is that page queries are allowed only on page components and static queries are allowed in every component. Static queries replace the graphql queries with json imports which create network requests.

I really like the idea that we are able to use graphql queries on the component level because it is super clean, but I'm buffled that it creates a dozen extra network requests. These requests are loaded on initial pageload.

In our usecase we use staticQueries mainly for images and we'd like to avoid prop drilling from the page component, since it is much cleaner to use the queries next to the lower level component.

Is there something I'm missing or understand differently? Is there a best practice to avoid the network requests from bloating?

Best,
Adam

**Comment by vladar** [comment]
Good question. Originally static queries were added to support use cases where you need to fetch data in the layout. So that you have maybe 3 to 5 static queries per site.

We realize that people use static queries for many things now, including images. And as an alternative I suggest you try the new `gatsby-plugin-image` beta: https://github.com/gatsbyjs/gatsby/discussions/27950

It has a separate `StaticImage` component that you can use with static images without GraphQL (and maybe static URLs in the near future)

**Reply by herecydev** [reply]
@vladar Your answer only helps with images. What solution do we have for static queries for things like localization, i.e. each component needs some labels it gets from graphql.

**Reply by harrygreen** [reply]
In addition:
- In most cases, the content is usually already in the HTML, so doesn't need fetching again on the client.
- They're added as `<link as="fetch" rel="preload">`, which is telling the browser they're "High" priority.

In our case, static queries offer very little other than a better development experience (no prop-drilling). Our solution is to move these static queries back into the page-level queries, which are only run on the server. 

Bandwidth usage:

![image](https://user-images.githubusercontent.com/1975098/126531773-ba62d87b-ec95-43b6-8ae7-14d700caca8e.png)

Static query JSON requests:

![image](https://user-images.githubusercontent.com/1975098/126531738-ac0190ff-9368-472b-bf87-bfbf627c6630.png)

**Reply by herecydev** [reply]
@harrygreen I'm thinking of doing the same. Did you see any observable differences? With HTTP/2 it negates a lot of the impact as it's not "waterfalling" but I can't help feel there's still a cost throwing many files around 🤔

**Reply by harrygreen** [reply]
@herecydev I'll post back here with my findings. I expect that moving the static queries to the page-level query will result in a bigger page-data.json.

**Reply by herecydev** [reply]
Oh that it definitely will, I'm just keen to see if loading 1 bigger file is quicker than 100 smaller files. If it is, then I can put forward a proposal to automatically merge the static queries into the page data.

**Comment by StevenFrostwick** [comment]
Hi @turbek 

Did you ever come up with a solution for this problem?

It would be great if all of the requests could be automatically batched up per page. Some of our pages are making requests for 30 or 40 JSON files.

A lot of them contain one or two translation strings, which we originally load from JSON files, so maybe there's a better solution for those which could resolve this problem.

**Reply by turbek** [reply]
Hi @StevenFrostwick

We needed internationalization support and the existing libraries weren’t really suitable to our usecase.

We create ‘content’ json files which we transform to graphql nodes and query these on the page level and filter it based on some meta data, e.g. ‘language’. We are drilling the content of these queried graphql nodes through the components. 

We call the createPage API on every language.

So effectively we decided to drop the staticQueries.

**Reply by StevenFrostwick** [reply]
Hi @turbek 

Yes that's how we have it set up as well, we're now in the process of "pulling" all the queries up to the page level so we don't have so many requests for these JSON files.

It's a shame because having a small `useStaticQuery` in the relevant components seems so much nicer from a code perspective than passing translations down through multiple components.

**Reply by turbek** [reply]
Indeed! I'm rooting for @herecydev in the other thread to go forward with a merging proposal:)
